src/auth/models.cpp

```diff
@@ -604,7 +604,14 @@ User::User(const std::string &username, std::optional<HashedPassword> password_h
 #endif
 
 bool User::CheckPassword(const std::string &password) {
-  return password_hash_ ? password_hash_->VerifyPassword(password) : true;
+  if (!password_hash_ && !username_.empty()) {
```

Comment:
Can it happen that `!password_hash_` evaluates to true and `!username_.empty()` to false? In that case `password_hash_->VerifyPassword(password)` will be called and that could lead to unexpected behaviors since password_hash_ isn't set.
If that can't happen then maybe `!username_.empty()` isn't needed here.